bin/varnishd/hash/hash_classic.c

```diff
@@ -171,13 +171,16 @@ hcl_lookup(struct worker *wrk, const void *digest, struct objhead **noh)
  * Dereference and if no references are left, free.
  */
 
-static int v_matchproto_(hash_deref_f)
-hcl_deref(struct objhead *oh)
+static int v_matchproto_(hash_deref_unlock_f)
+hcl_deref_unlock(struct worker *wrk, struct objhead *oh)
 {
 	struct hcl_hd *hp;
 	int ret;
 
 	CHECK_OBJ_NOTNULL(oh, OBJHEAD_MAGIC);
+	Lck_AssertHeld(&oh->mtx);
+	Lck_Unlock(&oh->mtx);
```

Comment:
I think this unlock is too early, we would at least need to do the oh->refcnt under the oh mtx

Reply:
The classic and simple hashes differ from the critbit in that they have a single global mutex that protects the refcounts of all the objheads, while critbit uses the objhead's own mutex to protect it's refcount. So I believe the code is correct as is.

Reply:
you are right.